spring-cloud-gcp-pubsub/src/main/java/org/springframework/cloud/gcp/pubsub/support/AcknowledgeablePubsubMessage.java

```diff
@@ -16,67 +16,21 @@
 
 package org.springframework.cloud.gcp.pubsub.support;
 
-import com.google.cloud.pubsub.v1.stub.SubscriberStub;
-import com.google.pubsub.v1.AcknowledgeRequest;
-import com.google.pubsub.v1.ModifyAckDeadlineRequest;
+import com.google.cloud.pubsub.v1.AckReplyConsumer;
 import com.google.pubsub.v1.PubsubMessage;
 
 /**
  * A {@link PubsubMessage} wrapper that allows it to be acknowledged.
  *
  * <p>To acknowledge {@link AcknowledgeablePubsubMessage} in bulk, using a
  * {@link org.springframework.cloud.gcp.pubsub.core.subscriber.PubSubSubscriberOperations#ack(java.util.Collection)}
- * is recommended. To do that, a user must first extract the ack IDs and group them by subscription name.
+ * is recommended.
  *
  * @author João André Martins
+ * @author Mike Eltsufin
  */
-public class AcknowledgeablePubsubMessage {
+public interface AcknowledgeablePubsubMessage extends AckReplyConsumer {
 
-	private PubsubMessage message;
-
-	private String ackId;
-
-	private String subscriptionName;
-
-	private SubscriberStub subscriberStub;
-
-	public AcknowledgeablePubsubMessage(PubsubMessage message, String ackId,
-			String subscriptionName, SubscriberStub subscriberStub) {
-		this.message = message;
-		this.ackId = ackId;
-		this.subscriptionName = subscriptionName;
-		this.subscriberStub = subscriberStub;
-	}
-
-	public PubsubMessage getMessage() {
-		return this.message;
-	}
-
-	public String getAckId() {
-		return this.ackId;
-	}
-
-	public String getSubscriptionName() {
-		return this.subscriptionName;
-	}
-
-	public void ack() {
-		AcknowledgeRequest acknowledgeRequest = AcknowledgeRequest.newBuilder()
-				.addAckIds(this.ackId)
-				.setSubscription(this.subscriptionName)
-				.build();
-
-		this.subscriberStub.acknowledgeCallable().call(acknowledgeRequest);
-	}
-
-	public void nack() {
-		ModifyAckDeadlineRequest modifyAckDeadlineRequest = ModifyAckDeadlineRequest.newBuilder()
-				.setAckDeadlineSeconds(0)
-				.addAckIds(this.ackId)
-				.setSubscription(this.subscriptionName)
-				.build();
-
-		this.subscriberStub.modifyAckDeadlineCallable().call(modifyAckDeadlineRequest);
-	}
+	PubsubMessage getPubsubMessage();
```

Comment:
Thoughts on adding `getSubscriptionName()` ?  I can see a scenario where code processing a message may want to know which subscription the message is from.

Reply:
Also what about the scenario where a developer wants to modify a messages ack deadline?

This would necessitate a `public void modifyAckDeadline(int ackDeadlineSeconds)` type of method in the interface.

Reply:
Adding `subscriptionName` should be OK.
Adding `modifyAckDeadline` would be an issue in the subscribe case because all we have there is the `AckReplyConsumer`, but can be add it to `PulledAcknowledgeablePubsubMessage`.

Reply:
You are correct, `modifyAckDeadline` would only make sense for a `PulledAcknowledgeablePubsubMessage`

Reply:
I would expose an API like `nack(ModifyAckDeadlineRequest)`. Looks like it is a pretty simple builder and can cover all your requirements.

Reply:
I don't think it's very user-friendly to expose `ModifyAckDeadlineRequest`. All we need is the seconds.

Reply:
M-m-m. Not sure what you mean...
Why does then Google do its API not user-friendly?..

This `ModifyAckDeadlineRequest` is really a part of Pub/Sub ack'ing API.
I'm sure there is nothing wrong to use it directly, without any Spring and that is really not a Spring goal to disable all the access to low-level client.
The simple sample is `JdbcTemplate` : it doesn't hide from us a fact of the `DataSource` fact. See `public DataSource getDataSource()`.

Something similar I have in the Spring Integration AWS, where alongside with our Spring API we are good to allow end-user to use existing AWS SDK abstractions.
There is just no reason to re-invent the wheel if existing one can be simply reused.

I don't mind if that not a case here though...

Feel free to do whatever you want and think is reasonable...

Reply:
I think the low-level API route is using `SubscriberFactory` that is exposed via the `template.getSubscriberFactory().getSubscriberStub()...`.

Reply:
OK. Let's go ahead with the solution and ship it to the field for feedback from target users!

No sense baulk any more! 😄